go/vt/vtgate/planbuilder/operators/insert.go

```diff
@@ -111,12 +111,245 @@ func createOperatorFromInsert(ctx *plancontext.PlanningContext, ins *sqlparser.I
 		return nil, err
 	}
 
-	vindexTable, routing, err := buildVindexTableForDML(ctx, tableInfo, qt, "insert")
+	vTbl, routing, err := buildVindexTableForDML(ctx, tableInfo, qt, "insert")
 	if err != nil {
 		return nil, err
 	}
 
-	insOp, err := createInsertOperator(ctx, ins, vindexTable, routing)
+	deleteBeforeInsert := false
+	if ins.Action == sqlparser.ReplaceAct &&
+		(ctx.SemTable.ForeignKeysPresent() || vTbl.Keyspace.Sharded) &&
+		(len(vTbl.PrimaryKey) > 0 || len(vTbl.UniqueKeys) > 0) {
+		// this needs a delete before insert as there can be row clash which needs to be deleted first.
+		ins.Action = sqlparser.InsertAct
+		deleteBeforeInsert = true
+	}
+
+	insOp, err := checkAndCreateInsertOperator(ctx, ins, vTbl, routing)
+	if err != nil {
+		return nil, err
+	}
+
+	if !deleteBeforeInsert {
+		return insOp, nil
+	}
+
+	rows, isRows := ins.Rows.(sqlparser.Values)
+	if !isRows {
+		return nil, vterrors.VT12001("REPLACE INTO using select statement")
+	}
+
+	pkCompExpr, err := pkCompExpression(vTbl, ins, rows)
+	if err != nil {
+		return nil, err
+	}
+
+	uniqKeyCompExprs, err := uniqKeyCompExpressions(vTbl, ins, rows)
+	if err != nil {
+		return nil, err
+	}
+
+	whereExpr := getWhereCondExpr(append(uniqKeyCompExprs, pkCompExpr))
+
+	delStmt := &sqlparser.Delete{
+		TableExprs: sqlparser.TableExprs{sqlparser.CloneRefOfAliasedTableExpr(ins.Table)},
+		Where:      sqlparser.NewWhere(sqlparser.WhereClause, whereExpr),
+	}
+	delOp, err := createOpFromStmt(ctx, delStmt, false, "")
+	if err != nil {
+		return nil, err
+	}
+	return &Sequential{Sources: []ops.Operator{delOp, insOp}}, nil
+}
+
+func getWhereCondExpr(compExprs []*sqlparser.ComparisonExpr) sqlparser.Expr {
+	var outputExpr sqlparser.Expr
+	for _, expr := range compExprs {
+		if expr == nil {
+			continue
+		}
+		if outputExpr == nil {
+			outputExpr = expr
+			continue
+		}
+		outputExpr = &sqlparser.OrExpr{
+			Left:  outputExpr,
+			Right: expr,
+		}
+	}
+	return outputExpr
+}
+
```

Comment:
this method is big and confusing enough to be improved by a comment explaining it's purpose